_posts/2020-10-13-uni-retry.adoc

```diff
@@ -0,0 +1,170 @@
+---
+layout: post
+title: Mutiny - How does retry... retries?
+date: 2020-10-13
+tags: reactive mutiny retry
+synopsis: How does `retry` actually work?
+author: cescoffier
+---
+
+Last week, David, a Quarkus user, asked me about retrying an asynchronous operation. 
+David has a _picky_ remote HTTP service, which sometimes misbehaves.
+Easy answer! 
+Just do: `uni.onFailure().retry().atMost(x)`. 
+But, David is curious, and asked me a second question: how does retry work? 
+Well, that’s simple; it just retries...
+As you can imagine, that did not satisfy David’s curiosity. 
+
+While I was answering to David, I realized that retrying is not that simple and deserves more explanation.
+That’s what we are going to see in this blog post.
+
+## Disclaimer about retries
+
+Okay, if you are the kind of reader who skips the terms and conditions, you can jump to the next section. 
+But, for others, I need to warn you about retries. 
+Retries may not be the solution to your problem, as it can be terrible.
+Retrying can only be done if your system can handle duplicated requests or messages.
+In other words, you can only retry if your system is idempotent, i.e., sending a request or a message multiple times does not change the overall state.
+In doubt, check before implementing a retry, as it can harm your system.
+
+Disclaimer said! Let’s look under the hood of retry.
+
+## Retry is about resubscribing
+
+Let’s imagine you have a `Uni` representing your asynchronous action, like in David’s case, an invocation of a remote service:
+
+[source, java]
+----
+Uni<String> uni = invokePickyService(client);
+----
+
+Unis are lazy.
+Until someone subscribes to them, nothing happens.
+In our case, the request is only sent to the remote service when someone subscribes to the `uni`.
+So to execute the request, we need to subscribe:
+
+[source, java]
+----
+Uni<String> uni = invokePickyService(client);
+uni.subscribe().with(
+    resp -> System.out.println("Success: " + resp),
+    failure -> System.out.println("Failed: " + failure.getMessage())
+);
+----
+
+In Quarkus,  most of the time, you return the `Uni`, and Quarkus subscribes to it. 
```

Comment:
Looks like a doubled space after the first comma?

Reply:
Wow... how can you see that!